### DISCUSSION TITLE
New gamplay concept: DJ/Battle Royale.

### DISCUSSION THREAD

**Discussion by Trung0246** [ppy/osu#28009] [Ideas]
Obviously inspired by [recent beatmap of the day concept](https://github.com/ppy/osu/discussions/27983).

# DJ/Single player

Initially I thought this could be fun with the following concept while brainstorming:

- Non-stop gameplay should be the main major concept, with some optional inserted break if desired.
- Automatic medley/compilation generation (either applying DT/HT automatically to match specified BPM and performing fading between each beatmap (and beatmap sections) transition) with random N beatmaps (can be set to finite number or set it to be infinity for "endless"-ish). BPM can behave like Wind Up/Wind Down randomly.
- Optional automatic effect applied while playing (like nightcore/daycore/sound-voltex-ish audio morph/etc.).
- Star count can be gradually increase/decrease.
- Stats like OD/AR/CS can be gradually increase/decrease.
- Pattern-modifying mods can be also enable/disable (mirror, random, etc).
- Fun mod randomly enable/disable.
- Should be compatible with all game mode if possible.

Everything should be as dynamic as possible to be able to apply during mid gameplay.

# Battle Royale/Multiplayer

DJ gameplay concept could be played with multiplayer, therefore making this battle-royale esque. I thought of a simple game rule, not sure if it viable but just a very bare concept:
- Introduce new actions: attack. This action can be triggered in the style of manual combo (like other rhythm games) as where each players gather "energy" when successfully hitting notes. There's potential different kind of "attack" action that may cost different energy amount:
	- Increase OD.
	- Increase AR.
	- Increase CS.
	- Increase HP.
	- Increase star.
	- Increase bpm.
	- Change pattern to harder version.
	- Fun mod force enable with random configuration.
- For each minute passed (can be configured to any N minutes mark), all player getting "attacked", which induce random random difficulty increase like above.
- Attack targeting mechanism could be random like Tetris 99.
- Combos can be used as currency to spend to reduce the difficulty and potentially make things easier.
	- This could be something else idk, just an initial thought on how to make things easier, debatable.
- Beatmap pick sequence should be deterministic.
	- As some player have faster BPM, they will load the beatmap in the sequence first compare with other players that have low BPM and will eventually reach the beatmap in the sequence.

Some potential winning condition includes:
- Miss count (lower is better, "lives", etc).
- Gain as much PP/score as possible.
	- For this condition, limiting the gameplay session to maybe N minutes.
- Prevent HP to ran out.
	- Which allows "unlimited" gameplay session duration, like any other battle royale.

# Problems

Potential major hurdle I could think of is beatmap downloading. This could be pretty much the entire reason this DJ/Battle Royale mode will not exist at all.

### DJ/Single Player
Maybe can be somewhat doable since the game could just load a beatmap that match criteria in the local beatmap DB.

### Battle Royale/Multiplayer
Unsure if on-the-fly beatmap download during gameplay session could be done (since I don't know much about server-side stuff). Another potential bottleneck is lag spike caused by downloading beatmaps.

Potential problem includes unintentionally DDoS-ing bancho with song download. I'm not a core dev so I have no ideas about this.

Attack trigger complexity and difficulty reduce counter action: I'm not the best at game design so this could be potential hurdle as player will have to split their attention span. Could be intentional. UI/UX design for this could be painful.

### Beatmap selection

Unfortunately I don't know of the way to fix a particular issue that is since each individual beatmap, inconsistent diff spike section pick is possible. This eventually could be another major hurdle.

Another thing is due to having to keep deterministic nature of beatmap selection, star change is potentially not possible for battle royale/multiplayer since this may cause players to have different beatmap picked.
- One potential mitigation I could think of is capping/snapping star change, multiple advance branch is possible for the same gameplay timeline (ex: if a player final star is 2-3, the game would load the easy branch, else maybe 4-5 then middle branch, otherwise 6+ for hard branch) while still keeping the deterministic requirement.

### Other technical problems

Maybe legacy compatibility with stable, implementation feasibility, and potentially other codebase limitation.

More workload is another potential issue, so there's that.

### How would this fit into osu!

Now that I think more about this, the entire thing could be a way to encourage players to play beatmaps they usually don't touch, and just another way to play the game. For multiplayer side, this could be seen as a way to actually let players having intense competition, more fun to stream, etc.

# Draft/Documentation

Overall this is just my personal thought. There's no draft/design documentation nonetheless since I pretty much just came up with this as shower thought. I also didn't open official discussion for this yet so maybe others can chime in.

**Comment by GamebP** [comment]
@trung0246: Probably way too stupid as an idea. 💀

**Comment by Hissymaster** [comment]
>Introduce new actions: attack. This action can be triggered in the style of manual combo (like other rhythm games) as where each players gather "energy" when successfully hitting notes. There's potential different kind of "attack" action that may cost different energy amount:
    Increase OD.
    Increase AR.
    Increase CS.
    Increase HP.
    Increase star.
    Increase bpm.
    Change pattern to harder version.
    Fun mod force enable with random configuration.
For each minute passed (can be configured to any N minutes mark), all player getting "attacked", which induce random random difficulty increase like above.
Combos can be used as currency to spend to reduce the difficulty and potentially make things easier.
    This could be something else idk, just an initial thought on how to make things easier, debatable.

You are definitely going to need to have breaks to prepare/prepare for attacks and/or defenses.

In addition, Your wording allows for these changes to happen mid-map, Of which, changing the difficulty is going to be quite difficult programmatically to do

> Attack targeting mechanism could be random like Tetris 99.

...So is the chosen attack also random? considering each one costs different amounts of energy, how is the attack picked?

**Reply by Trung0246** [reply]
>You are definitely going to need to have breaks to prepare/prepare for attacks and/or defenses. In addition, Your wording allows for these changes to happen mid-map, Of which, changing the difficulty is going to be quite difficult programmatically to do

Yes this is intended to change mid-map. Otherwise if it hard to do then forced break is a good solution.

> ...So is the chosen attack also random? considering each one costs different amounts of energy, how is the attack picked?

Both attack type and attack target could be random to reduce complexity. But I've prefer some mechanism to selectively choose attack type. Maybe a `charge attack bar` that when the energy charge past certain amount the attack type will be executed. And attack type can be upgrade the longer the game goes. This helps since player can still keep playing continuously.

Either case if it will be forced break then we probably don't need the charge bar system.
